src/dune_rules/expander.ml

```diff
@@ -119,12 +119,19 @@ let expand_version scope pform s =
     | None -> [ Value.String "" ]
     | Some s -> [ String s ]
   in
+  let project = Scope.project scope in
   match
     Package.Name.Map.find
-      (Dune_project.packages (Scope.project scope))
+      (Dune_project.packages project)
       (Package.Name.of_string s)
   with
   | Some p -> value_from_version p.version
+  | None when Dune_project.dune_version project < (2, 9) ->
+    User_error.raise
+      ~loc:(String_with_vars.Var.loc pform)
+      [ Pp.textf
+          "Package %S doesn't exist in the current project and installed \
```

Comment:
Small nitpick: this also applies to packages from other projects in the workspace, which wasn't supported before either. We also reach this in case of typo, and the message might be confusing in this case. Here is another attempt:

```suggestion
          "Package %S doesn't exist in the current project." ]
          ~hints:[ Pp.text "If you want to refer to an installed package, or more generally to a package from another project, you need at least (lang dune 2.9)." ]
```